### PR TITLE
Add pid support

### LINKED ISSUE #19: Extend PIDSubsystem instead of Subsystem.
The `Drivetrain` subsystem extends the `Subsystem` class by default. Change it to extend the `PIDSubsystem` class instead since it has cool PID functions in it.
Docs [HERE](https://docs.wpilib.org/he/stable/docs/software/commandbased/pid-subsystems-commands.html)

### PR DESCRIPTION
**Pull request by ofirlang**
Closes #19

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/DriveTrain.java [hunk 1]
```diff
@@ -9,15 +9,22 @@
 import edu.wpi.first.wpilibj.ADIS16448_IMU;
 import edu.wpi.first.wpilibj.drive.DifferentialDrive;
 import edu.wpi.first.wpilibj.motorcontrol.MotorControllerGroup;
+import edu.wpi.first.wpilibj2.command.PIDSubsystem;
 import edu.wpi.first.wpilibj2.command.SubsystemBase;
 import frc.robot.Constants.DriveTrainConstants;
+import java.beans.Encoder;
```
Comment (guy-av) [inline src/main/java/frc/robot/subsystems/DriveTrain.java:15]: Wrong import

File: src/main/java/frc/robot/subsystems/DriveTrain.java [hunk 2]
```diff
@@ -9,15 +9,22 @@
...
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.math.controller.SimpleMotorFeedforward;
+
+public class DriveTrain extends PIDSubsystem {
   private WPI_TalonFX m_leftFront;
   private WPI_TalonFX m_leftRear;
   private WPI_TalonFX m_rightFront;
   private WPI_TalonFX m_rightRear;
 
+  private final SimpleMotorFeedforward m_DriveTrainFeedforward;
```
Comment (guy-av) [inline src/main/java/frc/robot/subsystems/DriveTrain.java:26]:
```suggestion
  private final SimpleMotorFeedforward m_driveTrainFeedforward;
```

File: src/main/java/frc/robot/subsystems/DriveTrain.java [hunk 3]
```diff
@@ -96,4 +109,18 @@ public double getRightWheelsRPM() {
   public double getAverageWheelsRPM() {
     return (getLeftWheelsRPM() + getRightWheelsRPM()) / 2;
   }
+
+  @Override
+  public double getMeasurement(){
+    return m_leftFront.getSelectedSensorPosition();
```
Comment (guy-av) [inline src/main/java/frc/robot/subsystems/DriveTrain.java:115]: Should return the average measure distance by all the encoders. This method was already added by @VenusM3, `getDistance()`

File: src/main/java/frc/robot/subsystems/DriveTrain.java [hunk 4]
```diff
@@ -96,4 +109,18 @@ public double getRightWheelsRPM() {
...
+  }
+
+  @Override
+  public void useOutput(double output, double setpoint){
+    m_leftMotors.setVoltage(output + m_DriveTrainFeedforward.calculate(setpoint));
```
Comment (guy-av) [inline src/main/java/frc/robot/subsystems/DriveTrain.java:120]: Set the output only to left motors? What about the right ones?
Setting only one side will send the robot spinning uncontrollably

File: src/main/java/frc/robot/Constants.java [hunk 5]
```diff
@@ -14,6 +14,17 @@ public static class DriveTrainConstants {
...
+    // Feed Forward constants
+    public static final double kS = 0.0;
+    public static final double kV = 0.0;
+
+    // PID constants
+    public static final double kP = 0.0;
+    public static final double kI = 0.0;
+    public static final double kD = 0.0;
+
+    public static final double kDistanceToTarget = 0.0;
```
Comment (guy-av) [inline src/main/java/frc/robot/Constants.java:27]: Your distance to target is usually not a constant, it's value that changes as the game progresses.

File: src/main/java/frc/robot/subsystems/DriveTrain.java [hunk 6]
```diff
@@ -26,17 +33,23 @@ public class DriveTrain extends SubsystemBase {
...
+
     m_leftMotors = new MotorControllerGroup(m_leftRear, m_leftFront);
     m_rightMotors = new MotorControllerGroup(m_rightRear, m_rightFront);
 
     m_drive = new DifferentialDrive(m_leftMotors, m_rightMotors);
 
     m_imu = new ADIS16448_IMU();
+
+    getController().setTolerance(getSetpoint());
+    setSetpoint(DriveTrainConstants.kDistanceToTarget);
```
Comment (guy-av) [inline src/main/java/frc/robot/subsystems/DriveTrain.java:52]: Do not set a SetPoint inside the constructor, your SetPoint changes according to where you are on the field it's not constant.

For example, if I stand near the feeder, I will need to drive a whole lot more to the charge station than I would need if I were to stand near the grid.

**Review by guy-av — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Constants.java [hunk 7]
```diff
@@ -16,6 +16,17 @@ public static class DriveTrainConstants {
...
+    // Feed Forward constants
+    public static final double kS = 0.0;
+    public static final double kV = 0.0;
+
+    // PID constants
+    public static final double kP = 0.0;
+    public static final double kI = 0.0;
+    public static final double kD = 0.0;
+
+    public static final double kInitialSetPoint = 0.0;
```
Comment (neatteamsoftware) [inline src/main/java/frc/robot/Constants.java:29]: delete this

Comment (neatteamsoftware) [reply src/main/java/frc/robot/Constants.java:29]: add tolerance instead

File: src/main/java/frc/robot/subsystems/DriveTrain.java [hunk 8]
```diff
@@ -26,17 +31,22 @@ public class DriveTrain extends SubsystemBase {
...
+    m_driveTrainFeedforward = new SimpleMotorFeedforward(DriveTrainConstants.kS, DriveTrainConstants.kV);
+
     m_leftMotors = new MotorControllerGroup(m_leftRear, m_leftFront);
     m_rightMotors = new MotorControllerGroup(m_rightRear, m_rightFront);
 
     m_drive = new DifferentialDrive(m_leftMotors, m_rightMotors);
 
     m_imu = new ADIS16448_IMU();
+
+    getController().setTolerance(DriveTrainConstants.kInitialSetPoint);
```
Comment (neatteamsoftware) [inline src/main/java/frc/robot/subsystems/DriveTrain.java:49]: use actual tolerance

File: src/main/java/frc/robot/subsystems/DriveTrain.java [hunk 9]
```diff
@@ -96,4 +106,24 @@ public double getRightWheelsRPM() {
...
+    m_rightMotors.setVoltage(output + m_driveTrainFeedforward.calculate(setpoint));
+  }
+
+  public boolean atSetpoint(){
+    return getController().atSetpoint();
+  }
+
+  @Override
+  public void periodic(){
+    setSetpoint(DriveTrainConstants.kInitialSetPoint - getMeasurement());
```
Comment (neatteamsoftware) [inline src/main/java/frc/robot/subsystems/DriveTrain.java:127]: Remove this line, you do not want to change the setpoint every 20ms... You want it to be set by the autonomous command that will run.

**Review by neatteamsoftware — CHANGES_REQUESTED** [review]

**Review by neatteamsoftware — APPROVED** [review]

**Review by guy-av — APPROVED** [review]

**Pull request merged**